services/graph/pkg/service/v0/base.go

```diff
@@ -0,0 +1,692 @@
+	}
+	// We need to return an untyped nil here otherwise the error==nil check won't work
+	return nil
+}
+
+func (g BaseGraphService) removeUserShare(ctx context.Context, permissionID string) error {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+		return err
+	}
+
+	removeShareResp, err := gatewayClient.RemoveShare(ctx,
+		&collaboration.RemoveShareRequest{
+			Ref: &collaboration.ShareReference{
+				Spec: &collaboration.ShareReference_Id{
+					Id: &collaboration.ShareId{
+						OpaqueId: permissionID,
+					},
+				},
+			},
+		})
+
+	if errCode := errorcode.FromCS3Status(removeShareResp.GetStatus(), err); errCode != nil {
+		return *errCode
+	}
+	// We need to return an untyped nil here otherwise the error==nil check won't work
+	return nil
+}
+
+func (g BaseGraphService) removeSpacePermission(ctx context.Context, permissionID string, resourceId *storageprovider.ResourceId) error {
+	grantee, err := spacePermissionIdToCS3Grantee(permissionID)
+	if err != nil {
+		return err
+	}
+
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+		return err
+	}
+	removeShareResp, err := gatewayClient.RemoveShare(ctx, &collaboration.RemoveShareRequest{
+		Ref: &collaboration.ShareReference{
+			Spec: &collaboration.ShareReference_Key{
+				Key: &collaboration.ShareKey{
+					ResourceId: resourceId,
+					Grantee:    &grantee,
+				},
+			},
+		},
+	})
+
+	if errCode := errorcode.FromCS3Status(removeShareResp.GetStatus(), err); errCode != nil {
+		return *errCode
+	}
+	// We need to return an untyped nil here otherwise the error==nil check won't work
+	return nil
+}
+
+func (g BaseGraphService) getUserPermissionResourceID(ctx context.Context, permissionID string) (*storageprovider.ResourceId, error) {
+	share, err := g.getCS3UserShareByID(ctx, permissionID)
+	if err != nil {
+		return nil, err
+	}
+	return share.GetResourceId(), nil
+}
+
+func (g BaseGraphService) getCS3UserShareByID(ctx context.Context, permissionID string) (*collaboration.Share, error) {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+		return nil, err
+	}
+
+	getShareResp, err := gatewayClient.GetShare(ctx,
+		&collaboration.GetShareRequest{
+			Ref: &collaboration.ShareReference{
+				Spec: &collaboration.ShareReference_Id{
+					Id: &collaboration.ShareId{
+						OpaqueId: permissionID,
+					},
+				},
+			},
+		})
+	if errCode := errorcode.FromCS3Status(getShareResp.GetStatus(), err); errCode != nil {
+		return nil, *errCode
+	}
+	return getShareResp.GetShare(), nil
+}
+
+func (g BaseGraphService) getPermissionByID(ctx context.Context, permissionID string, itemID *storageprovider.ResourceId) (*libregraph.Permission, *storageprovider.ResourceId, error) {
+	publicShare, err := g.getCS3PublicShareByID(ctx, permissionID)
```

Comment:
I would recommend to use a switch here since the different cases are exclusive. Something like
```golang
switch {
case err == nil:
case IsSpaceRoot (itemID):
case errors.As(err, &errcode) && errcode.GetCode() == errorcode.ItemNotFound:
default:
    return nil, err
}